File: src/test/java/kanti/sl/StateLanguageImplTest.java
package kanti.sl;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class StateLanguageImplTest {

	private final StateLanguage sl = StateLanguage.builder().build();

	@Test
	public void parse() {
		String line = "User:fullClassName=STRING-kanti.sl.User,name=STRING-Jojo,age=INTEGER-1000";
		User user = (User) sl.parse(User.class, line);
		User expected = new User("Jojo", 1000);
		Assertions.assertEquals(
			expected,
			user
		);
	}

	@Test
	public void from() {
		User user = new User("Jojo", 1000);
		String line = sl.from(User.class, user);
		Assertions.assertEquals(
			"User:fullClassName=STRING-kanti.sl.User,name=STRING-Jojo,age=INTEGER-1000",
			line
		);
	}

}

File: src/test/kotlin/kanti/sl/StateLanguageImplTest.kt
package kanti.sl

import kanti.sl.arguments.MutableStateArgument
import kanti.sl.arguments.MutableStateArguments
import kanti.sl.arguments.StateArguments
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class StateLanguageImplTest {

	private val sl: StateLanguage = StateLanguage.builder()
		.registerConverter(User::class.java, UserObjectConverter)
		.build()

	@Test
	fun parse() {
		val user = User("Jojo", 1000)
		val stg = sl.from(user.javaClass, user)
		Assertions.assertEquals(
			"User:name=Jojo,age=1000",
			stg
		)
	}

	@Test
	fun from() {
		val line = "User:name=Jojo,age=1000"
		val parsed = sl.parse(User::class.java, line)
		Assertions.assertEquals(
			User("Jojo", 1000),
			parsed
		)
	}

}

private object UserObjectConverter : StateObjectConverter {

	override fun convert(arguments: MutableStateArguments, state: Any) {
		if (state !is User)
			return
		arguments.put( "name", state.name)
		arguments.put("age", state.age)
	}

	override fun convert(arguments: StateArguments): User {
		val name = arguments["name"]?.value as String? ?: throw IllegalStateException("No name")
		val age = arguments["age"]?.value as Int? ?: throw IllegalStateException("No age")
		return User(
			name = name,
			age = age
		)
	}

}

private data class User(
	val name: String,
	val age: Int
)